### ISSUE TITLE
Tests only work with leader

### ISSUE THREAD

**Issue by ineiti** [dedis/cothority#2002]
While working on #2001, some tests showed up that only work if the leader is contacted. This seems bad in most cases, and should be extended to also work for non-leaders.

As far as I can tell, often something like this happens:
1. `AddTransaction` is sent to a node, it returns as soon as this node processed the transaction
2. `GetProof` is sent to two nodes, the node who doesn't have the transaction is faster and returns an empty proof

All tests that use `byzcoin.Client.UseNode` are suspicious:

- [x] `byzcoin.TestClient_GetProof`
- [x] `byzcoin.TestClient_GetProofCorrupted`
- [x] `byzcoin.TestSecureDarc`
- [x] `byzcoin.TestService_SetConfigRosterReplace`
- [x] `byzcoin.TestService_SetConfigRosterReplace`
- [x] `byzcoin.Client.DownloadState`
- [x] `byzcoin.contracts.TestDeferred_ScenarioMultiInstructions_`
- [x] `byzcoin.contracts.TestDeferred_ScenarioMultipleSigners`
- [x] `byzcoin.contracts.TestDeferred_ScenarioUpdateConfig`
- [x] `byzcoin.contracts.TestDeferred_SimpleDelete`
- [x] `byzcoin.contracts.TestDeferred_PublicDelete`
- [x] `byzcoin.contracts.TestInsecureDarc`
- [x] `calypso.TestService_ReshareLTS_Different`
- [x] `calypso.TestClient_CreateLTS`
- [x] `calypso.TestClient_Calypso`
- [x] `calypso.TestService_ReshareLTS_OneMore`
- [x] `eventlog.Client`
- [x] `personhood.Tests`

The following is good:
- `byzcoin.simulation.SimulationService.Run`, because all transactions need to be sent to the same node, else the counters are not correct
- `TestContract_Write_Benchmark` - same reason: multiple transactions with same signer

**Comment by Gilthoniel** [comment]
I'm facing this issue while rebasing and it raises a question: in the scenario where you send a transaction (thus using the parallel stuff) and wait, the following GetProof that will return a valid proof will only be valid for the node that returns it. The only other node that you'll be sure it can be contacted is the leader but any other can be in a different state (e.g. processing the block).

The consequence is that using the parallel function seems dangerous in multiple situations and it makes me wonder if it should be used more carefully, even something to manually opt-in. This could be very confusing for clients if you get a valid proof but then an error for another call.

Any opinions ?

**Comment by kc1212** [comment]
I think this is a question of how smart should the clients be?

Prior to parallel, clients assume the `AddTransactionAndWait` and `GetProof` are synchronous in the sense that if the first action completed successfully then the second one will too. With the parallel client, the client cannot use the same assumption anymore. It will certainly make client code harder to write because the client may need to wait for the block on a non-leader node. But it take the load off the leader.

In summary, we need to decide if we want to only support simpler (synchronous) clients. Otherwise we need to "educate" the clients to assume asynchronous behaviour and write code to handle it.

**Comment by ineiti** [comment]
I second what @kc1212 writes - more generally, using a distributed ledger is about asynchronous behaviour. If you want to have an always-synchronous, always-consistent storage, then you need to use a database.

Now for how much we can take this asynchronicity off the shoulders of the clients is a good question:
- `UseNode(0)` per default: the leader gets all the load and the client has an https://en.wikipedia.org/wiki/ACID access to the ledger
- have parallel per default: the load is distributed, if a node fails, others will reply, but the client has to be aware that sometimes he needs to use `WaitProof`

At least for the CLIs and the web-app the default should be parallel, else they are unusable... I suppose the same goes for java.

**Comment by Gilthoniel** [comment]
Note that there is a difference between parallel and distributed. But anyway, I wonder if the only problem here is that we return proof in any case and the `SendProtobufParellel` just assumes the first reply even if it is a proof of non-existence. It might be interesting to be able to tell the function what is considered as a correct reply so that you can continue the parellel tries until you get a proof of existence (if any).

**Comment by ineiti** [comment]
Of course every call is different, with regard to `GetProof`, you have the following possibilities:

1. correct proof of absence: the instance is not there and never will be
2. flaky proof of absence: the instance is not there _yet_, but it will be, and already _might_ be on another node
3. correct instance of the current state
4. outdated instance with regard to a valid transaction you sent to a node but that hasn't been added to this node

Other calls will have other requirements and possibilities. So to see what should be given to `SendProtobufParallelUntilI'mHappy`, I propose to make a list of calls and what they can reasonably return.

**Comment by Gilthoniel** [comment]
I made a PR that allows use to use a custom decoder so that when getting a proof for instance, you can specify the latest known index by the client, and then the decoder will make sure the latest block from the proof is at least from that index. I think that's a bit easier that checking the different cases you stated.
